python_modules/dagster/dagster/_core/definitions/assets.py

```diff
@@ -1059,95 +1085,50 @@ def with_attributes(
 
         backfill_policy = check.opt_inst_param(backfill_policy, "backfill_policy", BackfillPolicy)
 
-        if group_names_by_key:
-            group_name_conflicts = [
-                asset_key
-                for asset_key in group_names_by_key
-                if asset_key in self.group_names_by_key
-                and self.group_names_by_key[asset_key] != DEFAULT_GROUP_NAME
-            ]
-            if group_name_conflicts:
-                raise DagsterInvalidDefinitionError(
-                    "Group name already exists on assets"
-                    f" {', '.join(asset_key.to_user_string() for asset_key in group_name_conflicts)}"
-                )
+        conflicts_by_attr_name: Dict[str, Set[AssetKey]] = defaultdict(set)
+        replaced_specs = []
 
-        replaced_group_names_by_key = {
-            output_asset_key_replacements.get(key, key): group_name
-            for key, group_name in self.group_names_by_key.items()
-        }
-
-        if freshness_policy:
-            freshness_policy_conflicts = (
-                self.freshness_policies_by_key.keys()
-                if isinstance(freshness_policy, FreshnessPolicy)
-                else (freshness_policy.keys() & self.freshness_policies_by_key.keys())
+        for key, spec in self._specs_by_key.items():
+            replace_dict = {}
+
+            def update_replace_dict_and_conflicts(
+                new_value: Union[Mapping[AssetKey, object], object],
+                attr_name: str,
+                default_value: object = None,
+            ) -> None:
+                if isinstance(new_value, Mapping):
+                    if key in new_value:
+                        replace_dict[attr_name] = new_value[key]
+                elif new_value:
+                    replace_dict[attr_name] = new_value
+
+                old_value = getattr(spec, attr_name)
+                if old_value and old_value != default_value and attr_name in replace_dict:
+                    conflicts_by_attr_name[attr_name].add(key)
+
```

Comment:
imo we should have two functions here instead of discriminating on type for `Mapping` versus having two different behaviors which are subtly different

Reply:
The reason I went with one function is that some inputs for `new_value` will actually have type `Union[Mapping[AssetKey, object], object]`.

E.g. the `freshness_policy` argument to `with_attributes` can be either a `FreshnessPolicy` or a `Mapping[AssetKey, FreshnessPolicy]]`. Ditto with `auto_materialize_policy`.

This means if we changed it to two functions, we'd still need that branching higher up.

So my gut is to keep it this way, but also fine changing it if, in light of the above, you still think we should.